BAW mesh writer

Crisspl opened the pull request.

Crisspl committed:
*.baw mesh writer

Crisspl committed:
Edit Meshloaders example

Added few lines here to test CBAWMeshWriter

Crisspl committed:
Lil renaming

Removed index-buffer blob type

Crisspl committed:
Some baw meshloader fixing and some docs

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -0,0 +1,269 @@
+
+		uint32_t CBAWMeshWriter::genHeaders(ICPUMesh * _mesh)
+		{
+			m_headers.clear();
```
what about a blob for the _mesh itself!?

Crisspl replied:
Considered as not needed since it's just collection of ptrs to mesh buffers. Plus bounding box, but this can be recalculated after adding all loaded mesh buffers.

devshgraphicsprogramming replied:
Sure, but think what happens once we put more than one IMesh in a file.

P.S. The bounding box can be quite expensive to recalculate for a skinned mesh.

devshgraphicsprogramming commented on source/Irrlicht/CBawFile.h:
```diff
@@ -45,6 +60,7 @@ namespace irr { namespace core
 		};
 
+		//! Meant to be used as pointer to blob's data
 		uint8_t data[1];
```
Maybe boot this and replace with a 'void* getData() {return this;}'

Crisspl replied:
Yea, better

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -0,0 +1,269 @@
+		void CBAWMeshWriter::exportAsBlob<ICPUMeshBuffer>(ICPUMeshBuffer* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			uint8_t* const data = (uint8_t*)std::malloc(m_headers[_headerIdx].blobSizeDecompr);
+			uint8_t* nearestFree = data;
```
All this stuff is nice, but not very maintanable.. maybe make a "struct MeshBufferBlobV1 : public Blob" which will construct itself from an ICPUMeshBuffer*

And also use the stack instead of heap! 
Making our exporter faster!

Crisspl replied:
Nice idea,  thanks

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -0,0 +1,269 @@
+		{
+			uint8_t* const data = (uint8_t*)std::malloc(m_headers[_headerIdx].blobSizeDecompr);
+
+			E_COMPONENTS_PER_ATTRIBUTE* cpa = (E_COMPONENTS_PER_ATTRIBUTE*)data;
```
All this stuff is nice, but not very maintanable.. maybe make a "struct MeshDataFormatDescBlobV1 : public Blob" which will construct itself from an IMeshDataFormatDesc<ICPUBuffer>*

And also use the stack instead of heap! 
Making our exporter faster!

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -0,0 +1,269 @@
+
+			uint64_t* idxBuf = (uint64_t*)((uint8_t*)attBufs + EVAI_COUNT * sizeof(uint64_t));
+
+			for (int i = 0; i < EVAI_COUNT; ++i)
```
Loop unrolling will help and happen automagically by the compiler... but!

The memory is being walked very erratically, which the compiler won't fix, its much better for performance to have many loops for each array to be saved.

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -0,0 +1,269 @@
+					bh.handle = reinterpret_cast<uint64_t>(meshBuffer);
+					bh.compressionType = core::Blob::EBCT_RAW;
+					bh.blobType = core::Blob::EBT_MESH_BUFFER;
+					bh.blobSize = bh.blobSizeDecompr = // size of significant members of ICPUMeshBuffer
```
This is complicated to maintain, maybe we should set blobSizeDecompr later?

Or at least as sizeof(MeshBufferBlobV1) ?

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -0,0 +1,269 @@
+					bh.handle = reinterpret_cast<uint64_t>(desc);
+					bh.compressionType = core::Blob::EBCT_RAW;
+					bh.blobType = core::Blob::EBT_DATA_FORMAT_DESC;
+					bh.blobSize = bh.blobSizeDecompr = // size of members of ICPUMeshDataFormatDesc
```
This is complicated to maintain, maybe we should set blobSizeDecompr later?

Or at least as sizeof(IMeshDataFormatDescBlobV1) ?

Crisspl committed:
[wip] Changes after CR

Done: optimizations, blob for mesh itself, some new blob-structs, increased readability
TODO: Thread safety (will go in next commit because I'll need special CR on it)

Crisspl committed:
CBAWMeshWriter: Thread safety

Also fixed texture file path bug

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -28,40 +28,34 @@ namespace irr {
-			nearestFree += sizeof(E_VERTEX_ATTRIBUTE_ID);
-
+			const uint32_t cnt = _obj->getMeshBufferCount();
+			const uint32_t size = sizeof(uint32_t) + sizeof(core::aabbox3df) + cnt*sizeof(uint64_t);
```
Use placement new to create a MeshBlob over a larger data memory allocation

devshgraphicsprogramming commented on source/Irrlicht/CBAWMeshWriter.cpp:
```diff
@@ -28,40 +28,34 @@ namespace irr {
 		template<>
+		void CBAWMeshWriter::exportAsBlob<ICPUMeshBuffer>(ICPUMeshBuffer* _obj, uint32_t _headerIdx, io::IWriteFile* _file)
+		{
+			const core::MeshBufferBlobV1 data(*_obj);
```
ADT violation

devshgraphicsprogramming commented on source/Irrlicht/CBawFile.cpp:
```diff
@@ -0,0 +1,46 @@
+namespace irr { namespace core
+{
+
+	MeshBufferBlobV1::MeshBufferBlobV1(const scene::ICPUMeshBuffer& _mb)
```
change to pointer

Crisspl committed:
Final touches
